### PR TITLE
[SPARK-48314][SS] Don't double cache files for FileStreamSource using Trigger.AvailableNow

### PR DESCRIPTION
**Pull request by Kimahriman**
### What changes were proposed in this pull request?

Files don't need to be cached for reuse in `FileStreamSource` when using `Trigger.AvailableNow` because all files are already cached for the lifetime of the query in `allFilesForTriggerAvailableNow`.

### Why are the changes needed?

As reported in https://issues.apache.org/jira/browse/SPARK-44924 (with a PR to address https://github.com/apache/spark/pull/45362), the hard coded cap of 10k files being cached can cause problems when using a maxFilesPerTrigger > 10k. It causes every other batch to be 10k files, which can greatly limit the throughput of a new streaming trying to catch up.

### Does this PR introduce _any_ user-facing change?

Every other streaming batch won't be 10k files if using Trigger.AvailableNow and maxFilesPerTrigger greater than 10k.

### How was this patch tested?

New UT

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by Kimahriman** [comment]
@HeartSaVioR since you added the file caching originally back in the day

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 1]
```diff
@@ -2357,6 +2357,46 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
...
+        CountListingLocalFileSystem.resetCount()
+
+        // provide 20 files in src, with sequential "last modified" to guarantee ordering
+        (0 to 19).map { idx =>
+          val f = createFile(idx.toString, new File(src, idx.toString), tmp)
+          f.setLastModified(idx * 10000)
+          f
+        }
+
+        source.latestOffset(FileStreamSourceOffset(-1L), ReadLimit.maxFiles(5))
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:2379]: Shall we check the result just for completeness sake?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 2]
```diff
@@ -2357,6 +2357,46 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
...
+        // All files are already tracked in allFilesForTriggerAvailableNow
+        assert(0 === CountListingLocalFileSystem.pathToNumListStatusCalled
+          .get(src.getCanonicalPath).map(_.get()).getOrElse(0))
+
+        // Unread files should be empty
+        assert(source.unreadFiles == null)
+
+        // Reading again leverages the files already tracked in allFilesForTriggerAvailableNow,
+        // so no more listings need to happen
+        source.latestOffset(FileStreamSourceOffset(-1L), ReadLimit.maxFiles(5))
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:2391]: Same here.

**Review by HeartSaVioR — COMMENTED** [review]
Looks good in overall. Left a few comments for better testing.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala (hunk 2 above)
Comment (Kimahriman) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Same as above

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**